src/MLJInterface.jl

```diff
@@ -0,0 +1,163 @@
+module MLJInterfaceModule
+
+using MLJModelInterface: MLJModelInterface
+import DynamicExpressions: eval_tree_array, string_tree, Node
+import ..CoreModule: Options, Dataset, LOSS_TYPE
+import ..ComplexityModule: compute_complexity
+import ..HallOfFameModule: HallOfFame, calculate_pareto_frontier, format_hall_of_fame
+#! format: off
+import ..equation_search
+#! format: on
+
+mutable struct SRRegressor <: MLJModelInterface.Deterministic
+    sr_options::Options
+    niterations::Int
+    parallelism::Symbol
+    numprocs::Union{Int,Nothing}
+    procs::Union{Vector{Int},Nothing}
+    addprocs_function::Union{Function,Nothing}
+    runtests::Bool
+    loss_type::Type
+    selection_method::Function
+end
+# TODO: Ideally we could forward all parameters from `Options`?
+
+function choose_best(; trees, losses::Vector{L}, scores, complexities) where {L<:LOSS_TYPE}
+    # Same as in PySR:
+    # https://github.com/MilesCranmer/PySR/blob/e74b8ad46b163c799908b3aa4d851cf8457c79ef/pysr/sr.py#L2318-L2332
+    # threshold = 1.5 * minimum_loss
+    # Then, we get max score of those below the threshold.
+    threshold = 1.5 * minimum(losses)
+    return argmax([
+        (losses[i] <= threshold) ? scores[i] : typemin(L) for i in eachindex(losses)
+    ])
+end
+
+function dispatch_selection(
+    selection_method::F, trees::VN, losses, scores, complexities
+) where {F,T,N<:Node{T},VN<:Vector{N}}
+    return selection_method(;
+        trees=trees, losses=losses, scores=scores, complexities=complexities
+    )::Integer
+end
+function dispatch_selection(
+    selection_method::F, trees::MN, losses, scores, complexities
+) where {F,T,N<:Node{T},VN<:Vector{N},MN<:Vector{VN}}
+    return [
+        dispatch_selection(
+            selection_method, trees[i], losses[i], scores[i], complexities[i]
+        ) for i in eachindex(trees)
+    ]
+end
+
+# TODO: Should somehow automatically document kws available.
+function SRRegressor(;
```

Comment:
The function keyword arguments should correspond to the struct keys.

Reply:
The dataset is documented here: https://astroautomata.com/SymbolicRegression.jl/stable/types/#SymbolicRegression.CoreModule.DatasetModule.Dataset. The user normally does not need to worry about this though as it is for internal use, for storing the variable names, units, normalization, etc.

Reply:
The dataset is useful if the user wants to define a complex custom objective; they can pass any other data their function needs to the `extra` parameter (3D tensors, graphs, etc). But this is uncommon and probably incompatible with other ML algorithms.

Reply:
> This is especially important for model hyperparameter tuning

I see, thanks. So one option is for us to generate the SRRegressor struct using the keywords available to the `Options` constructor. And then, inside `fit`, the normal `Options` can be created to be passed to `equation_search`. (Some keywords in the `Options` constructor are different from the stored keys)